File: src/main/java/com/hatchways/blogposts/service/PostServiceImpl.java
package com.hatchways.blogposts.service;

import com.hatchways.blogposts.model.Post;
import com.hatchways.blogposts.model.User;
import com.hatchways.blogposts.repository.PostRepository;
import com.hatchways.blogposts.repository.UserRepository;
import com.hatchways.blogposts.schema.PostRequest;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import org.modelmapper.ModelMapper;

@org.springframework.stereotype.Service
public class PostServiceImpl implements PostService {
  private final PostRepository postRepository;
  private final UserRepository userRepository;
  private final ModelMapper modelMapper;

  public PostServiceImpl(
      PostRepository postRepository, UserRepository userRepository, ModelMapper modelMapper) {
    this.postRepository = postRepository;
    this.userRepository = userRepository;
    this.modelMapper = modelMapper;
  }

  @Override
  public Post createPost(PostRequest postRequestBody, String username) {
    Post post = modelMapper.map(postRequestBody, Post.class);
    Set<User> users = new HashSet<>();
    User user = userRepository.findByUsername(username);
    users.add(user);
    post.setUsers(users);
    postRepository.save(post);
    postRepository.flush();
    return post;
  }

  @Override
  public List<Post> fetchPosts() {
    try {
      return this.postRepository.findAll();
    } catch (Exception ex) {
      throw new RuntimeException("Internal Server Error, please contact our support team");
    }
   
  }
}

File: src/main/java/com/hatchways/blogposts/service/PostServiceImpl.kt
package com.hatchways.blogposts.service

import com.hatchways.blogposts.exception.BadRequestException
import com.hatchways.blogposts.model.Post
import com.hatchways.blogposts.model.User
import com.hatchways.blogposts.repository.PostRepository
import com.hatchways.blogposts.repository.UserRepository
import com.hatchways.blogposts.schema.PostRequest
import org.modelmapper.ModelMapper
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service

@Service
class PostServiceImpl(
        private val postRepository: PostRepository,
        private val userRepository: UserRepository,
        private val modelMapper: ModelMapper
) : PostService {

  override fun createPost(postRequestBody: PostRequest, username: String): Post {
    val post = modelMapper.map(postRequestBody, Post::class.java)
    val user = userRepository.findByUsername(username)
      ?: throw BadRequestException("User not found")
    post.users = setOf(user)
    postRepository.save(post)
    postRepository.flush()
    return post
  }

  override fun fetchPosts(pageSize: Int, pageNumber: Int, sortingOrder: String): Page<Post> {
    val sortOrder = if (sortingOrder == "ASC") Sort.Direction.ASC else Sort.Direction.DESC
    val sort = Sort.by(sortOrder, "id")
    val pageable: Pageable = PageRequest.of(pageNumber, pageSize, sort)

    val page = postRepository.findAll(pageable)
    if (pageNumber > page.totalPages) {
      throw BadRequestException("Page number exceeds total pages")
    }
    return page
  }
}